feat(aws-lambda) Lambda authentication via EC2 IAM roles

Mgutjahr opened the pull request:
### Summary
Adds support for querying temporary credentials from EC2 metadata service instead of fetching long-lived key-pairs from the database.

The credentials are fetched from the EC2 metadata service and are cached for 60 seconds. I assume that many users are running kong on EC2 instances. For these users this is the recommended way to authenticate requests for AWS services as key-rotation and credential provisioning are managed by AWS.

For further details refer to the [AWS documentation about EC2 roles](http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html)

### Full changelog

* Added `kong.plugins.aws-lambda.iam-role-credentials` for  fetching and caching temporary credentials from the metadata service (default 169.254.169.254 port 80) 
* Added plugins config flag `use_ec2_iam_role` to `kong.plugins.aws-lambda.schema` and declared `aws_key` and `aws_secret` as optional
* Added `self_check` function to check whether either `use_ec2_iam_role` is set  to true or both aws_key and aws_secret as configured
* Added if branch in `kong.plugins.aws-lambda.handler` to either fetch credentials from the metadata service or use credentials stored in database

p0pr0ck5 commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,91 @@
+local http = require 'resty.http'
+local json = require 'cjson'
+local cache = require "kong.tools.database_cache"
```
minor style comment: please align assignment operators here

p0pr0ck5 commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,91 @@
+    metadata_service_port = metadata_service_port or 80
+ 
+    local client = http.new()
+    local ok, err = client:connect(metadata_service_host, metadata_service_port) 
```
style: please align assignment operators here

p0pr0ck5 commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,91 @@
+    end 
+     
+    if role_name_request_res.status ~= 200 then
+      return nil, "[aws-lambda] Fetching role name from metadata service returned status code ".. role_name_request_res.status  .. 
```
style: can we try to adjust line breaks and such so as to avoid the line being much longer than 80 chars

p0pr0ck5 commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,91 @@
+
+    local iam_security_token_data = json.decode(iam_security_token_request:read_body())
+    
+    ngx.log(ngx.INFO, "[aws-lambda] Received temporary IAM credential from metadata service for role '" ..
```
this also seems like it should be a debug level log entry

thibaultcha commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,91 @@
+    local ok, err = client:connect(metadata_service_host, metadata_service_port) 
+    
+    if not ok then
+      ngx.log(ngx.ERR, "Could not connect to metadata service: " .. err)
```
@Mgutjahr Please avoid Lua-land string concatenation, and use the variadic arguments capabilities of `ngx.log` instead:
```
ngx.log(ngx.ERR, "[aws-lambda] could not connect to metadata service: ", err)
```

Mgutjahr commented:
Hi @p0pr0ck5, 
thanks for the detailed feedback and for pointing out references to possible solutions! 

I addressed most of your recommendations and fixed the mentioned style issues. On the testing side I added a tests for the schema extension. In addition, I added an integration test which loads credentials from a mocked-out metadata service (using the same approach aws-lambda uses) 

I decided not to add support for configuring the role the plugin should load as one EC2 instance can only have role configured at a time. This can be found from the [AWS documentation:](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html)
> An instance profile can contain only one IAM role. This limit cannot be increased.

This is also how almost all of the available AWS SDK's decided to implement support for instance roles. 

Sorry again for the style issues and the missing tests.

Mgutjahr commented:
Hi @p0pr0ck5, 
I fixed the requested style issues. Hope it looks better now. Good call on the testing front, we should definitely add a test for a non-existing metadata service as this could be a possible error scenario. 

On creating the test I realised that the metadata service call had no connection timeout configured, as this could eat up resources I added a 500ms timeout for the metadata service calls (this should be sufficient as the metadata service runs locally on EC2 instances. Requests usually take ~1ms).

thibaultcha commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,83 @@
+
+local function fetch_iam_credentials_from_metadata_service(metadata_service_host, metadata_service_port)
+    local client = http.new()
+    client:set_timeout(500)
```
It's dangerous to make such assumptions considering Kong might be running in a variety of infrastructures with different limitations or requirements out there. Such timeout variables should ideally be made configurable.

p0pr0ck5 replied:
Not quite sure about the validity of this comment, given the very narrow infrastructure scope of the context at play here. Adding a tunable for this seems like unnecessary cognitive load.

thibaultcha replied:
Our "mantra" of being opinionated (in order to not oversight any potential use-case or to be compatible with as many infrastructures as possible), hints that we should provide customization timeouts. In fact there shouldn't be a single timeout (to the best of my knowledge) that isn't configurable in Kong already, including this module itself.

p0pr0ck5 replied:
I think the idea here is that inter-infrastructure compatibility is meaningless here. Is it worth sticking to a hard-line mantra in such a case? IMO it's not, but this isn't a hill for me to die on, just chiming in that this is a very specific use case :)

Mgutjahr replied:
I get your concerns, I tried to omit any unnecessary config properties to keep the plugin configuration  relatively clean. Especially as the timeout would be another semi-optional config property (only applies if role based authentication is used). From my experience (using native AWS sdks on a variety of platforms) this is a property you would never need to touch. But maybe there are cases where you have too. 

I checked other "official" SDK's from Amazon. Most set a default timeout which cannot be changed. 
- [python cli](https://github.com/aws/aws-cli/blob/e2295b022db35eea9fec7e6c5540d06dbd6e588b/awscli/customizations/codedeploy/utils.py#L113) --> 1 second
- [python boto](https://github.com/boto/botocore/blob/cac78632cabddbc7b64f63d99d419fe16917e09b/botocore/session.py#L96) --> 1 second (can be configured)
- [java sdk](https://github.com/aws/aws-sdk-java/blob/7844c64cf248aed889811bf2e871ad6b276a89ca/aws-java-sdk-core/src/main/java/com/amazonaws/internal/ConnectionUtils.java#L41) --> 2 seconds

The timeouts from these SDK's are higher than the one set by me, but most of timeouts mentioned above apply for all services offered by the metadata service (and not just the credentials service). 

I'd suggest that I increase the timeout to 5 seconds by default and expose it as a parameter to `get_iam_credentials_from_instance_profile()`. - This way we can tweak it for the test and exposing it to the plugins config afterwards would be trivial if needed. 

If you want me to add it to the config I can do this of course too. Just let me know what you'd prefer.

thibaultcha commented on kong/plugins/aws-lambda/iam-role-credentials.lua:
```diff
@@ -0,0 +1,83 @@
+    }
+     
+    if not role_name_request_res or role_name_request_res == "" then
+          ngx.log(ngx.ERR, "[aws-lambda] Could not fetch role name from metadata service: ", err)
```
(Some weird indentation here again)

thibaultcha commented on kong/plugins/aws-lambda/schema.lua:
```diff
@@ -20,11 +22,13 @@ return {
       type = "string",
-      required = true,
+    },
+    use_ec2_iam_role = {
```
Our policy is that no plugin config already in the database should be left in a previous version of the plugin's schema. This is because we want to avoid complicated code paths handling different versions of the same plugin's schema in the plugin business logic itself. This is a policy that has unfortunately not always been followed, and such oversight lead us to a few bugs/overhead maintenance issues recently.

So: new schema value, new migration for Cassandra/PostgreSQL :)

Mgutjahr replied:
Hi @thibaultcha, 
are you referring to the `use_ec2_iam_role` config property introduced in this PR? - The aws_secret and aws_key were changed to semi-optional, so whatever config was in the database is still correct after the PR (it will continue to use secrets & keys from the db config if `use_ec2_iam_role` is not set / set to false). In other words, as long as use_ec2_iam_role is not set to true the old schema is still enforced. 

Do you mean I should configure `use_ec2_iam_role` to a required property and set it to false by default through a migration script? I'm unsure of the benefits of this. Wouldn't it be better to remove the default=false flag here?

thibaultcha replied:
Hmm indeed... I do agree with that reasoning. We kind of set this "unsaid" policy of enforcing migrations when introducing new values, but optional values would not benefit from it in any way, because they might be `nil` per definition already.

Alright then, this sounds good! Sorry for the noise!

p0pr0ck5 commented:
@Mgutjahr can you rebase this commit to fix the conflict with the rockspec? After that I think we're good to go here.

Mgutjahr commented:
Hi @p0pr0ck5, 
sorry for my late reply. I rebased the branch to the current master branch. Note that I had to move the caching of IAM credentials to the handler due to the change of the caching in 0.11.

Mgutjahr commented:
Hi @thibaultcha, do you have any updates/comments on this PR?

p0pr0ck5 commented:
@Mgutjahr sorry for the delay on our end. Apologies, since we've had another release since the last change, can we update the rockspec here? I'd like to see this merged into 0.12. @thibaultcha thoughts?

fyntic commented:
Hello @p0pr0ck5 , @thibaultcha  and @Mgutjahr .

Sorry for disturb you, guys, but we really need this feature available in Kong for AWS.
When do you think it might be merged and released?

Thanks in advance.

aloisbarreras commented:
Is all that needs to happen just rebasing the rockspec? I'm happy to do that real fast if that will help this get merged in.

vvondra commented:
is it ok if i re-open this fixed?

farazs commented:
@thibaultcha is all that's left for this to update the rockspec and resolve the handler.lua conflict? seems something trivial and we can merge this.

@Mgutjahr would you be able to do that or someone else can fork this and make the necessary updates. would be really nice to get this merged since it's so close and quite useful.

tarciosaraiva commented:
Any chance of this getting reviewed soon?

Mgutjahr commented:
Hey, I rebased the PR again to the current master branch. The branch could be merged if you want to. There are just two things I'd like to point out: 
* The IAM role authentication currently **only works if kong is running on an EC2 instance**. It would not work if Kong is running on ECS, Fargate or EKS. We can add support for ECS and Fargate but I would not do this in the same PR. Inspiration on how to do this can be found in the various official AWS SDK's out there (e.g. [Java](https://docs.amazonaws.cn/AWSJavaScriptSDK/latest/AWS/RemoteCredentials.html) or [Node](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/remote_credentials.js) ) .. there is unfortunately no real documentation out there from Amazon (at least I could not find it) 
* This PR is also covered in PR #3639. I'm ok if you close this PR and continue on #3639, the PR mentioned has however additional features & functionality we might want to discuss in more detail. Especially proxying of the metadata service could be a bit risky from a security perspective (as you get IAM credentials through this service and proxying could increase the attach surface for a man in the middle attack)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=2777) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=2777) before we can accept your contribution.<br/><hr/>**Manuel Gutjahr** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=2777) it.</sub>
